agent/collector/http_server.go

```diff
@@ -0,0 +1,204 @@
+	jsonContentType     = "application/json"
+)
+
+type httpServer struct {
+	addr     string
+	ingester ingester
+
+	hServer *http.Server
+}
+
+func newHttpServer(addr string, ingester ingester) *httpServer {
+	return &httpServer{
+		addr:     addr,
+		ingester: ingester,
+	}
+}
+
+func (s *httpServer) Start() error {
+	r := mux.NewRouter()
+	r.HandleFunc("/v1/traces", s.Export).Methods("POST")
+
+	s.hServer = &http.Server{
+		Addr:    s.addr,
+		Handler: handlers.CompressHandler(decompressBodyHandler(handlers.ContentTypeHandler(r, protoBufContentType, jsonContentType))),
+	}
+	listener, err := net.Listen("tcp", s.addr)
+	if err != nil {
+		return fmt.Errorf("cannot listen on address %s: %w", s.addr, err)
+	}
+
+	return s.hServer.Serve(listener)
+}
+
+func (s *httpServer) Stop() {
+	s.hServer.Close()
+}
+
+func (s httpServer) Export(w http.ResponseWriter, r *http.Request) {
+	contentType := r.Header.Get("content-type")
+	response := newHttpResponse(w, contentType)
+
+	request, err := s.parseBody(r.Body, contentType)
+	if err != nil {
+		response.sendError(http.StatusUnprocessableEntity, status.Errorf(codes.InvalidArgument, "Could not parse request body %s", err.Error()))
+		return
+	}
+
+	result, err := s.ingester.Ingest(r.Context(), request, "HTTP")
+	if err != nil {
+		response.sendError(http.StatusInternalServerError, status.Errorf(codes.InvalidArgument, "Error when ingesting spans %s", err.Error()))
+		return
+	}
+
+	response.send(http.StatusOK, result)
+}
+
+func (s httpServer) parseProtoBuf(body []byte) (*pb.ExportTraceServiceRequest, error) {
+	request := pb.ExportTraceServiceRequest{}
+
+	err := proto.Unmarshal(body, &request)
+	if err != nil {
+		return nil, err
+	}
+
+	return &request, nil
+}
+
+func (s httpServer) parseJson(body []byte) (*pb.ExportTraceServiceRequest, error) {
+	exportRequest := ptraceotlp.NewRequest()
+
+	err := exportRequest.UnmarshalJSON(body)
+	if err != nil {
+		return nil, err
+	}
+
+	protoBody, err := exportRequest.MarshalProto()
+	if err != nil {
+		return nil, err
+	}
+
+	return s.parseProtoBuf(protoBody)
+}
+
+func (s httpServer) parseBody(reqBody io.ReadCloser, contentType string) (*pb.ExportTraceServiceRequest, error) {
+	var body []byte
+	if b, err := io.ReadAll(reqBody); err == nil {
+		body = b
+	} else {
+		return nil, err
+	}
+
+	if len(body) == 0 {
+		return nil, fmt.Errorf("empty body")
+	}
+
+	if contentType == protoBufContentType {
+		return s.parseProtoBuf(body)
+	}
+
+	return s.parseJson(body)
+}
+
+type httpResponse struct {
+	w           http.ResponseWriter
+	contentType string
+}
+
+func newHttpResponse(w http.ResponseWriter, contentType string) httpResponse {
+	return httpResponse{
+		w:           w,
+		contentType: contentType,
+	}
+}
+
+func (r httpResponse) send(statusCode int, message proto.Message) error {
+	body, err := r.paseResponseBody(message)
+	if err != nil {
+		fmt.Println("Could not attach body to response", err.Error())
+		return err
```

Comment:
Does it make sense to wrap this error with this message too?
```suggestion
		fmt.Println("Could not attach body to response", err.Error())
		return fmt.Errorf("could not attach body to response: %w", err)
```